src/TopOptEMFocus.jl

```diff
@@ -0,0 +1,550 @@
+# ```
+# 
+# where $u_{x/y}$ is the depth into the PML, $\sigma$ is a profile function (here we chose $\sigma(u)=\sigma_0(u/d_{pml})^2$) and different derivative corresponds to different absorption directions.  Note that at a finite mesh resolution, PML reflects some waves, and the standard technique to mitigate this is to "turn on" the PML absorption gradually—in this case we use a quadratic profile. The amplitude $\sigma_0$ is chosen so that in the limit of infinite resolution the "round-trip" normal-incidence is some small number. 
+# 
+# Since PML absorbs all waves in $x/y$ direction, the associated boundary condition is then usually the zero Dirichlet boundary condition. Here, the boundary conditions are zero Dirichlet boundary on the top and bottom side $\Gamma_D$ but periodic boundary condition on the left ($\Gamma_L$) and right side ($\Gamma_R$).  The reason that we use a periodic boundary condition for the left and right side instead of zero Dirichlet boundary condition is that we want to simulate a plane wave exicitation, which then requires a periodic boundary condition.
+# 
+# Consider $\mu(x)=1$ (which is mostly the case in electromagnetic problems) and denote $\Lambda=\operatorname{diagm}(\Lambda_x,\Lambda_y)$ where $\Lambda_{x/y}=\frac{1}{1+\mathrm{i}\sigma(u_{x/y})/\omega}$, we can formulate the problem as 
+# 
+# ```math
+# \left\{ \begin{aligned} 
+# \left[-\Lambda\nabla\cdot\frac{1}{\varepsilon(x)}\Lambda\nabla -k^2\right] H &= f(x) & \text{ in } \Omega,\\
+# H&=0 & \text{ on } \Gamma_D,\\
+# H|_{\Gamma_L}&=H|_{\Gamma_R},&\\
+# \end{aligned}\right.
+# ```
+# 
+# For convenience, in the weak form and Julia implementation below we represent $\Lambda$ as a vector instead of a diagonal $2 \times 2$ matrix, in which case $\Lambda\nabla$ becomes the elementwise product.
+# 
+# ## Topology Optimization
+# 
+# We use the density-based topology optimizaiton (TO) to optimize the electric field intensity at center. In TO, every point in the design domain is a design degree of freedom, which we discretize into a piece-wise constant parameter space $P$ for the design parameter $p\in [0,1]$. The material electric permittivity can then be determined by
+# 
+# ```math
+# \varepsilon(p) = \left[n_{air}+p(n_{metal}-n_{air})\right]^2, 
+# ```
+# where $n_{air}=1$ and $n_{metal}$ is the refractive index of the air and metal. Note that we do not directly interpolate the electric permittivity in the design region using the electric permittivity of air and metal, because the metal permittivity is negative and the direct interpolate might yield some divergence [4]. 
+# 
+# In practice, to avoid obtaining arbitrarily fine features as the spatial resolution is increased, one needs to impose a non-strict minimum lengthscale $r_f$. There are two schemes for this smoothing filter [5], here we perform the smoothing by solving a simple "damped diffusion" PDE, also called a Helmholtz filter [5]: 
```

Comment:
```suggestion
# In practice, to avoid obtaining arbitrarily fine features as the spatial resolution is increased, one needs to regularize the problem with a minimum lengthscale $r_f$ by generating a smoothed/filtered parameter function $p_f$.  (Although this regularizes the problem, strictly speaking it does not impose a minimum feature size because of the nonlinear-projection step below. In practical applications, one imposes additional [manufacturing constraints](http://doi.org/10.1364/OE.431188) explicitly.)  We perform the smoothing $p \to p_f$ by solving a simple "damped diffusion" PDE, also called a Helmholtz filter [5], for $p_f$ given the design variables $p$: 
```